### PR TITLE
BUG: Restore return value in notebook reprs

### LINKED ISSUE #10491: Output on objects defining _ipython_display_
I have a usage question / maybe API enhancement request about `_ipython_display_`. I noticed today that objects defining a `_ipython_display_` method don't seem to be able to trigger the `Out[n]` machinery for the output of a cell. [Here's an example notebook](http://nbviewer.jupyter.org/gist/TomAugspurger/081d6ae057fb2f856c8398400fd12be7) Notice that the `In[5]` doesn't have an `Out[5]`.

Is there any way in `_ipython_display_` to say "treat this one as the return value". Should IPython treat the highest-priority one as the return value?

This came up in a pandas issue https://github.com/pandas-dev/pandas/issues/16168#issuecomment-298077926

### PR DESCRIPTION
**Pull request by TomAugspurger**
Monkey patches the _ipython_display_ method onto NDFrame, so that
notebook cells have a real return value. Setting the
display.html.table_schema will monkey patch the method on,
and remove it when unset.

closes https://github.com/ipython/ipython/issues/10491

### CODE REVIEW

File: pandas/core/config_init.py [hunk 1]
```diff
@@ -307,6 +307,48 @@ def mpl_style_cb(key):
     return val
 
 
+def table_schema_cb(key):
+    # Having _ipython_display_ defined messes with the return value
+    # from cells, so the Out[x] dicitonary breaks.
```
Comment (jreback) [inline pandas/core/config_init.py]: don't write the full method here
instead inside the cb import a method a call it

Comment (jreback) [inline pandas/core/config_init.py]: actually i would create this dynamically
simply define 2 methods and have the cb just set one of them

we do in set_use_numexpr just like this

Comment (jreback) [inline pandas/core/config_init.py]: would not create dynamically (typo)

File: pandas/core/generic.py [hunk 2]
```diff
@@ -6283,6 +6258,38 @@ def logical_func(self, axis=None, bool_only=None, skipna=None, level=None,
...
+    # monkey patch `_ipython_display_` onto NDFrame when config option
+    # is set
+    # see https://github.com/pandas-dev/pandas/issues/16168
+    try:
+        from IPython.display import display
+    except ImportError:
+        return None
+
+    # Series doesn't define _repr_html_ or _repr_latex_
+    latex = self._repr_latex_() if hasattr(self, '_repr_latex_') else None
```
Comment (jreback) [inline pandas/core/generic.py]: maybe we should define his for Series (trivial version) to avoid this kind of logic

Comment (TomAugspurger) [inline pandas/core/generic.py]: https://github.com/pandas-dev/pandas/issues/16180

**Comment by TomAugspurger** [comment]
@jreback look ok to you?

**Comment by jreback** [comment]
yep

**Pull request merged**